specs/light_client/merkle_proofs.md

```diff
@@ -64,43 +59,134 @@ y_data_root  len(y)
 We can now define a concept of a "path", a way of describing a function that takes as input an SSZ object and outputs some specific (possibly deeply nested) member. For example, `foo -> foo.x` is a path, as are `foo -> len(foo.y)` and `foo -> foo.y[5].w`. We'll describe paths as lists, which can have two representations. In "human-readable form", they are `["x"]`, `["y", "__len__"]` and `["y", 5, "w"]` respectively. In "encoded form", they are lists of `uint64` values, in these cases (assuming the fields of `foo` in order are `x` then `y`, and `w` is the first field of `y[i]`) `[0]`, `[1, 2**64-1]`, `[1, 5, 0]`.
 
 ```python
-def path_to_encoded_form(obj: Any, path: List[Union[str, int]]) -> List[int]:
-    if len(path) == 0:
-        return []
-    elif isinstance(path[0], "__len__"):
-        assert len(path) == 1
-        return [LENGTH_FLAG]
-    elif isinstance(path[0], str) and hasattr(obj, "fields"):
-        return [list(obj.fields.keys()).index(path[0])] + path_to_encoded_form(getattr(obj, path[0]), path[1:])
-    elif isinstance(obj, (Vector, List)):
-        return [path[0]] + path_to_encoded_form(obj[path[0]], path[1:])
+def item_length(typ: SSZType) -> int:
+    """
+    Returns the number of bytes in a basic type, or 32 (a full hash) for compound types.
+    """
+    if issubclass(typ, BasicValue):
+        return typ.byte_len
+    else:
+        return 32
+        
+        
+def get_elem_type(typ: ComplexType, index: Union[int, str]) -> Type:
+    """
+    Returns the type of the element of an object of the given type with the given index
+    or member variable name (eg. `7` for `x[7]`, `"foo"` for `x.foo`)
+    """
+    return typ.get_fields()[index] if issubclass(typ, Container) else typ.elem_type        
+
+
+def chunk_count(typ: SSZType) -> int:
+    """
+    Returns the number of hashes needed to represent the top-level elements in the given type
+    (eg. `x.foo` or `x[7]` but not `x[7].bar` or `x.foo.baz`). In all cases except lists/vectors
+    of basic types, this is simply the number of top-level elements, as each element gets one
+    hash. For lists/vectors of basic types, it is often fewer because multiple basic elements
+    can be packed into one 32-byte chunk.
+    """
+    # typ.length describes the limit for list types, or the length for vector types.
+    if issubclass(typ, BasicValue):
+        return 1
+    elif issubclass(typ, Bits):
+        return (typ.length + 255) // 256
+    elif issubclass(typ, Elements):
+        return (typ.length * item_length(typ.elem_type) + 31) // 32
+    elif issubclass(typ, Container):
+        return len(typ.get_fields())
     else:
-        raise Exception("Unknown type / path")
+        raise Exception(f"Type not supported: {typ}")
+
+
+def get_item_position(typ: SSZType, index: Union[int, str]) -> Tuple[int, int, int]:
```

Comment:
Worried about the `[int, str]` a bit, since they are exclusively (from each other) used based on `typ`. So e.g. Container with int is invalid. It is likely something to revisit with partials/other uses, but is ok for now.

Reply:
I think in all cases where we expect to use paths, the path would not be serialized, and it would be explicitly constructed directly as an input to `get_generalized_index` or similar methods (ie. never saved as a variable). So I think it's fine for that reason.

Reply:
Yes agree there, but we can maybe make it less error prone. We have others that are unfamiliar with the interface and need to use it as well.